test/mapping/irregular_grid.jl

```diff
@@ -47,3 +48,365 @@ basins, attractors = basins_of_attraction(newton)
+
+# ds = DiscreteDynamicalSystem(newton_map, [0.1, 0.2], [3.0])
+# xg = yg = range(-1.5, 1.5, length = 100)
+
+
+# #xg = yg = range(0, 5; length = 100)
+
+# grid = (xg, yg)
+
+# using LinearAlgebra
+
+# function make_irregular_array(ds, grid, maxlevel = 4)
+#     indices = CartesianIndices(length.(grid))
+#     f, p = dynamic_rule(ds), current_parameters(ds)
+#     udummy = copy(current_state(ds))
+#     #velocities = Dict()
+#     velocities = zeros(length.(grid))
+
+#     for ind in indices
+#         u0 = Attractors.generate_ic_on_grid(grid, ind)
+#         velocity = if !isinplace(ds)
+#             f(u0, p, 0.0)
+#         else
+#             f(udummy, u0, p, 0.0)
+#             udummy
+#         end
+#         if (isequal(LinearAlgebra.norm(velocity),NaN))
+#             #velos[ind] = 1
+#             velocities[ind] = 1
+#         else 
+#             #velos[ind] = LinearAlgebra.norm(velocity)
+#             velocities[ind] = LinearAlgebra.norm(velocity)
+#             if LinearAlgebra.norm(velocity)>100
+#                 println("error")
+#             end
+#         end
+#     end
+#     #println(velos)
+    
+
+
+#     #println(velocities)
+#     maxvel = maximum(velocities)
+#     #println(maxvel)
+#     #println(velocities[maxvel])
+#     velratios = maxvel./velocities
+
+#     result = [round(Int,log2(clamp(x, 1, 2^maxlevel))) for x in velratios]
+#     #println(velratios)
+#     #velratios = Dict(x => maxvel[2]/velocities[x] for x in keys(velocities))
+#     #result = Dict(x =>  round(Int,log2(clamp(velratios[x], 1, 2^maxlevel))) for x in keys(velratios))
+
+#     return result
+# end
+# Set(make_irregular_array(ds, grid))
+
+
+
+
+# function make_irregular_array(ds, grid, maxlevel = 4)
+#     indices = CartesianIndices(length.(grid))
+#     f, p = dynamic_rule(ds), current_parameters(ds)
+#     udummy = copy(current_state(ds))
+#     velocities = Dict()
+#     #velocities = zeros(length.(grid))
+
+#     for ind in indices
+#         u0 = Attractors.generate_ic_on_grid(grid, ind)
+#         velocity = if !isinplace(ds)
+#             f(u0, p, 0.0)
+#         else
+#             f(udummy, u0, p, 0.0)
+#             udummy
+#         end
+#         if (isequal(LinearAlgebra.norm(velocity),NaN))
+#             #velos[ind] = 1
+#             velocities[ind] = 1
+#         else 
+#             #velos[ind] = LinearAlgebra.norm(velocity)
+#             velocities[ind] = LinearAlgebra.norm(velocity)
+#             if LinearAlgebra.norm(velocity) > 100
+#                 println("error 2")
+#             end
+
+#         end
+#     end
+#     #println(velos)
+    
+
+
+#     maxvel = maximum(values(velocities))
+#     println(maxvel)
+#     velratios = Dict(x => maxvel/velocities[x] for x in keys(velocities))
+#     println(velratios)
+#     result = Dict(x =>  round(Int,log2(clamp(velratios[x], 1, 2^maxlevel))) for x in keys(velratios))
+
+#     return result
+# end
+# println(Set(values(make_irregular_array(ds, grid))))
+
+
+
```

Comment:
What's going on here, where are all the tests? When we had the video call you showed me amazing tests where you picked individual points, made a Subdivision based grid, and then showed that the `basin_cell_index` correctly gives you the points you want. Please add **four characteristic examples** of picking a subdivision array, and then selecting a point, and showing that it returns the correct index. Please also attach the Excel file you have created to test this.

Reply:
I am sorry, that is not even close to the  final version. That just one which works. I am cleaning up the code now and providing tests and documentation. Just intermediate commit.

Reply:
Aaaaah okay, sorry. I submited a review as if this was finished for review :D